packages/core/test/ifthenelse_test.js

```diff
@@ -0,0 +1,424 @@
+          },
+          then: {
+            properties: { telephone: { type: "string" } },
+          },
+        },
+      ],
+    };
+
+    it("should render correctly when condition is true in allOf (1)", () => {
+      const formData = {
+        country: "United States of America",
+      };
+
+      const { node } = createFormComponent({
+        schema: schemaWithAllOf,
+        formData,
+      });
+
+      expect(node.querySelector("input[label=zipcode]")).not.eql(null);
+    });
+
+    it("should render correctly when condition is false in allOf (1)", () => {
+      const formData = {
+        country: "",
+      };
+
+      const { node } = createFormComponent({
+        schema: schemaWithAllOf,
+        formData,
+      });
+
+      expect(node.querySelector("input[label=zipcode]")).to.eql(null);
+    });
+
+    it("should render correctly when condition is true in allof (2)", () => {
+      const formData = {
+        country: "United Kingdom",
+      };
+
+      const { node } = createFormComponent({
+        schema: schemaWithAllOf,
+        formData,
+      });
+
+      expect(node.querySelector("input[label=postcode]")).not.eql(null);
+      expect(node.querySelector("input[label=zipcode]")).to.eql(null);
+      expect(node.querySelector("input[label=telephone]")).to.eql(null);
+    });
+
+    it("should render correctly when condition is true in allof (3)", () => {
+      const formData = {
+        country: "France",
+      };
+
+      const { node } = createFormComponent({
+        schema: schemaWithAllOf,
+        formData,
+      });
+
+      expect(node.querySelector("input[label=postcode]")).to.eql(null);
+      expect(node.querySelector("input[label=zipcode]")).to.eql(null);
+      expect(node.querySelector("input[label=telephone]")).not.eql(null);
+    });
+
+    const schemaWithAllOfRef = {
+      type: "object",
+      properties: {
+        street_address: {
+          type: "string",
+        },
+        country: {
+          enum: [
+            "United States of America",
+            "Canada",
+            "United Kingdom",
+            "France",
+          ],
+        },
+      },
+      definitions: {
+        unitedkingdom: {
+          properties: { postcode: { type: "string" } },
+        },
+      },
+      allOf: [
+        {
+          if: {
+            properties: { country: { const: "United Kingdom" } },
+          },
+          then: {
+            $ref: "#/definitions/unitedkingdom",
+          },
+        },
+      ],
+    };
+
+    it("should render correctly when condition is true when then contains a reference", () => {
+      const formData = {
+        country: "United Kingdom",
+      };
+
+      const { node } = createFormComponent({
+        schema: schemaWithAllOfRef,
+        formData,
+      });
+
+      expect(node.querySelector("input[label=postcode]")).not.eql(null);
+    });
+  });
+
+  it("handles nested if then else", () => {
```

Comment:
#1666 didn't handle this case properly. A small change in the `resolveSchema` function fixed it (changing `if "if" in schema` to `while "if" in schema`), and here's a test to prove it.